todo_app/app.py

```diff
@@ -1,92 +1,223 @@
+            user_id_to_update = form_data["id"]
+            new_role = form_data["new_role"]
+            mongo_client.update_user(user_id_to_update, new_role)
+      else:
+         print("Reader Role is not allowed to start a task")   
+      return redirect('/users')
+
    @app.route('/items/restart', methods = ['POST'])
+   @login_required
    def restart_item():
-      if request.method == 'POST':
-         form_data = request.form
-         task_id = form_data["id"]
-         mongo_client.reopen_task(task_id)
-         
+      if (disable_login or current_user.role == writer_role):
+         if request.method == 'POST':
+               form_data = request.form
+               task_id = form_data["id"]
+               mongo_client.reopen_task(task_id)
+      else:
+         print("Reader Role is not allowed to restart a task")     
       return redirect('/')
 
+   @app.route('/login/callback', methods = ['GET'])
+   def login():
+      code = request.args.get("code")
+      
+      if request.method == 'GET':
+         code = request.args.get("code")
+         # Prepare and send a request to get tokens
+         token_url, headers, body = client.prepare_token_request(
+            "https://github.com/login/oauth/access_token",
+            authorization_response=request.url,
+            redirect_url=request.base_url,
+            code=code  
+         )
+
+         headers['Accept'] = 'application/json'
+
+         token_response = requests.post(
+            token_url,
+            headers=headers,
+            data=body,
+            auth=(oauth_client_id, oauth_secret_id)
+         )
+         
+         
+         client.parse_request_body_response(json.dumps(token_response.json()))
+         # access token is now stored in the client.access_token param
+
+         userinfo_endpoint = "https://api.github.com/user"
+         uri, headers, body = client.add_token(userinfo_endpoint)
+         userinfo_response = requests.get(uri, headers=headers, data=body)
+         
+         login_id = userinfo_response.json()["login"]
+         
+         # see if we can find the user in the database already
+         user = mongo_client.get_user_by_username(login_id)
+         if user is None:
+            # new users get added with Reader rights
+            user = mongo_client.add_user(login_id, reader_role)           
+         
+         # Begin user session by logging the user in
+         print(f"Logging in {user.username} in to the session")
+         session['username'] = user.username
+         login_user(user)
+         return redirect('/')
 
-   if __name__ == '__main__':
-      app.run()
 
 
+   if __name__ == '__main__':
+      app.run(ssl_context="adhoc")
+ 
+   @login_manager.unauthorized_handler 
+   def unauthenticated(): 
+      authorize_endpoint = 'https://github.com/login/oauth/authorize'
+      
+      authorize_url = client.prepare_request_uri(
+        authorize_endpoint, "http://127.0.0.1:5000/login/callback")
```

Comment:
Note that your callback URL will only work for localhost if it's set up like this. The easiest solution is to remove that parameter entirely and just relying on configuring the callback URL via Github, or you could allow this to be passed through by Environment variable (but it will still have to match Github's registered callback address!)